module/validation/receipt_validator.go

```diff
@@ -160,64 +161,203 @@ func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
 		if found {
 			return prevResult, nil
 		}
 
-		return v.previousResult(executionResult)
+		return v.fetchResult(previousResultID)
 	}
 
 	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+		prevResult, err := fetchResult(r.ExecutionResult.PreviousResultID)
```

Comment:
maybe we can introduce a `map[<result_id>]struct{}` to avoid validating the same result more than once, in the the block has 2 receipts having the same result.

Reply:
as you said in previous comment, it's a left over when we had method for validating multiple receipts at the same time, I will change it so it validates single receipt and that should solve both of your comments